### DISCUSSION TITLE
Netbox Inventory for objects other than Devices

### DISCUSSION THREAD

**Discussion by jof** [netbox-community/ansible_modules#1140] [Ideas]
I am using Netbox in a somewhat unusual setting for an Internet Exchange (IX), where we operate a shared Layer 2 Ethernet service, but have connections from devices that are outside of our administrative control.

We assign some IP addresses to IX participants which are tracked as IP Addresses in Netbox, with the participant-ownership tracked as a "Tenant" field on the IP address. However, these IP addresses are never assigned to device interfaces within our administrative control. But, we do wish to reference these Netbox objects inside of our Ansible when generating our DNS zone files and Route Server configurations.

Looking at the `netbox.netbox.nb_inventory` inventory plugin, it seems that it is only possible to query a list of Devices and objects that are associated with those Devices.

This prompts a couple questions from me:
- Should this inventory plugin be renamed in a way to make it clear that is is meant for querying Devices?
- Would there be interest from maintainers in accepting contributions of alternative inventory plugins that would support querying other Netbox Object types? (For example, IP Addresses)

As much as possible, I would like to avoid creating a custom, one-off inventory plugin.

**Comment by jof** [comment]
Since Ansible's inventory is inherently host-oriented, maybe it does make some sense that the inventory plugin is primarily indexed by Devices.

Looking more into some possibilities, I might be able to work around this in a couple ways:

- Using `netbox.netbox.nb_lookup` with a local task and registering the resulting objects into variables.

- Using Ansible's `ansible.builtin.uri` module to POST GraphQL queries to netbox and registering the resulting structure as variables.

**Comment by sc68cal** [comment]
You should use `nb_lookup`. It's much more flexible

**Comment by jof** [comment]
Indeed, I think I will go forward using the `netbox.netbox.nb_lookup` plugin.

For now, I'm able to cover my IP address enumeration (using tags for object filtration) need with a task like:

```
---
- name: Pull all Participant data from Netbox
  delegate_to: localhost
  run_once: true
  ansible.builtin.set_fact:
    netbox_participant_ip_addresses: "{{ query(
      'netbox.netbox.nb_lookup',
      'ip-addresses',
      api_endpoint=ixp_netbox_api_endpoint,
      token=ixp_netbox_api_token,
      api_filter='tag=ixp_participant'
      ) | items2dict }}"
```
